sentry-ruby/lib/sentry/profiler.rb

```diff
@@ -0,0 +1,232 @@
+# frozen_string_literal: true
+
+return unless defined?(StackProf)
+
+require 'securerandom'
+
+module Sentry
+  class Profiler
+
+    VERSION = '1'
+    PLATFORM = 'ruby'
+    # 101 Hz in microseconds
+    DEFAULT_INTERVAL = 1e6 / 101
+
+    attr_reader :sampled, :started, :event_id
+
+    def initialize
+      @event_id = SecureRandom.uuid.delete('-')
+      @started = false
+      @sampled = nil
+    end
+
+    def start
+      return unless @sampled
+
+      @started = StackProf.start(interval: DEFAULT_INTERVAL,
+                                 mode: :wall,
+                                 raw: true,
+                                 aggregate: false)
+
+      @started ? log('Started') : log('Not started since running elsewhere')
+    end
+
+    def stop
+      return unless @sampled
+      return unless @started
+
+      StackProf.stop
+      log('Stopped')
+    end
+
+    def profile_context
+      { profile_id: @event_id }
+    end
+
+    # Sets initial sampling decision of the profile.
+    # @return [void]
+    def set_initial_sample_decision(transaction_sampled)
+      unless Sentry.initialized? && Sentry.configuration.profiling_enabled?
+        @sampled = false
+        return
+      end
+
+      unless transaction_sampled
+        @sampled = false
+        log('Discarding profile because transaction not sampled')
+        return
+      end
+
+      profiles_sample_rate = Sentry.configuration.profiles_sample_rate
+
+      case profiles_sample_rate
+      when 0.0
+        @sampled = false
+        log('Discarding profile because sample_rate is 0')
+        return
+      when 1.0
+        @sampled = true
+        return
+      else
+        @sampled = Random.rand < profiles_sample_rate
+      end
+
+      log('Discarding profile due to sampling decision') unless @sampled
+    end
+
+    def to_hash
+      return nil unless Sentry.initialized?
+      return nil unless @sampled
+      return nil unless @started
+
+      results = StackProf.results
+      return nil unless results
+      return nil if results.empty?
+      return nil if results[:samples] == 0
+      return nil unless results[:raw]
+
+      frame_map = {}
+
+      frames = results[:frames].to_enum.with_index.map do |frame, idx|
+        frame_id, frame_data = frame
+
+        # need to map over stackprof frame ids to ours
+        frame_map[frame_id] = idx
+
+        in_app = in_app?(frame_data[:file])
+        filename = compute_filename(frame_data[:file], in_app)
+        function, mod = split_module(frame_data[:name])
+
+        {
+          abs_path: frame_data[:file],
+          function: function,
+          module: mod,
+          lineno: frame_data[:line],
+          filename: filename,
+          in_app: in_app
+        }.compact
```

Comment:
TIL `Hash#compact` 👍 
But if looks like the only value could be `nil` here is `frame_data[:line]`? How often does that happen? If it's not a common case, we should probably change this to:


```rb
        hash = {
          abs_path: frame_data[:file],
          function: function,
          module: mod,
          filename: filename,
          in_app: in_app
        }
        hash[:lineno] = frame_data[:line] if frame_data[:line]
        hash
```

Otherwise, `Hash#compact` always allocate a new hash

```
irb(main):006:0> h = { a: 1 }; puts h.object_id; puts h.compact.object_id
291420
291440                                                                                                                      
=> nil    
```

which is probably not much, but IMO could be avoided here.

Reply:
good point, `module` can also be `nil`, but I will add both optionally.